### PR TITLE
Apr 1753 gateway migration from admin

### PR DESCRIPTION
**Pull request by David-B-Read**

### CODE REVIEW

File: src/SFA.DAS.RoatpGateway.Domain/CriminalCompliancePageConfiguration.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
+﻿using System.Collections.Generic;
+
+namespace SFA.DAS.RoatpGateway.Domain
+{
+    public static class CriminalCompliancePageConfiguration
+    {
+        public static Dictionary<string, string> Headings = new Dictionary<string, string>
```
Comment (martin-hornsby) [inline src/SFA.DAS.RoatpGateway.Domain/CriminalCompliancePageConfiguration.cs:7]: I think this needs to be a: private static IReadOnlyDictionary

File: src/SFA.DAS.RoatpGateway.Domain/PaginatedList.cs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        public abstract int TotalPages { get; }
+
+        public abstract bool HasPreviousPage { get; }
+        public abstract bool HasNextPage { get; }
+
+        public abstract int FirstVisiblePage { get; }
+        public abstract int LastVisiblePage { get; }
+
+        public abstract int FirstVisibleItem { get; }
+        public abstract int LastVisibleItem { get; }
```
Comment (martin-hornsby) [inline src/SFA.DAS.RoatpGateway.Domain/PaginatedList.cs:22]: Some of these properties are abstract but their deriving class doesn't provide any special implementation. Wouldn't it be better to have them as virtual ?

File: src/SFA.DAS.RoatpGateway.Domain/ProviderTypes.cs [hunk 3]
```diff
@@ -0,0 +1,10 @@
+﻿
+namespace SFA.DAS.RoatpGateway.Domain
+{
+    public static class ProviderTypes
```
Comment (martin-hornsby) [inline src/SFA.DAS.RoatpGateway.Domain/ProviderTypes.cs:4]: This looks like an enum to me

File: src/SFA.DAS.RoatpGateway.Domain/Roatp/OrganisationType.cs [hunk 4]
```diff
@@ -0,0 +1,11 @@
+﻿
+namespace SFA.DAS.RoatpGateway.Domain.Roatp
+{
+    public class OrganisationType
+    {
+        public int Id { get; set; }
+        public string Type { get; set; }
+        public string Description { get; set; }
+        public const int Unassigned = 0;
```
Comment (martin-hornsby) [inline src/SFA.DAS.RoatpGateway.Domain/Roatp/OrganisationType.cs:9]: What does Unassigned do? Looks a bit odd in a POCO class

File: src/SFA.DAS.RoatpGateway.Domain/RoatpGatewayConstants.cs [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+            {
+                public static string CompaniesHouse = "Companies House data";
+                public static string CharityCommission = "Charity Commission data";
+            }
+        }
+    }
+
+    public static class NoSelectionErrorMessages
+    {
+        public static Dictionary<string, string> Errors = new Dictionary<string, string>
```
Comment (martin-hornsby) [inline src/SFA.DAS.RoatpGateway.Domain/RoatpGatewayConstants.cs:60]: Shouldn't this be a: private static IReadOnlyDictionary

File: src/SFA.DAS.RoatpGateway.Web/Controllers/RoatpShutterPageController.cs [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+using Microsoft.AspNetCore.Mvc;
+using System.Threading.Tasks;
+
+namespace SFA.DAS.RoatpGateway.Web.Controllers
+{
+    [Authorize]
+    public class RoatpShutterPageController : Controller
+    {
+        [Route("ExternalApisUnavailable")]
+        public async Task<IActionResult> ExternalApisUnavailable()
```
Comment (martin-hornsby) [inline src/SFA.DAS.RoatpGateway.Web/Controllers/RoatpShutterPageController.cs:11]: This is not an async method

File: src/SFA.DAS.RoatpGateway.Web/Infrastructure/ApiClients/RoatpOrganisationSummaryApiClient.cs [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+using Microsoft.Extensions.Logging;
+using SFA.DAS.RoatpGateway.Web.Infrastructure.ApiClients.TokenService;
+using SFA.DAS.RoatpGateway.Web.Models;
+
+namespace SFA.DAS.RoatpGateway.Web.Infrastructure.ApiClients
+{
+    public class RoatpOrganisationSummaryApiClient : ApiClientBase<RoatpOrganisationSummaryApiClient>, IRoatpOrganisationSummaryApiClient
+    {
+        private readonly HttpClient _client;
+        private readonly ILogger<RoatpOrganisationSummaryApiClient> _logger;
```
Comment (martin-hornsby) [inline src/SFA.DAS.RoatpGateway.Web/Infrastructure/ApiClients/RoatpOrganisationSummaryApiClient.cs:15]: These are in ApiClientBase

File: src/SFA.DAS.RoatpGateway.Web/Controllers/RoatpGatewayRegisterChecksController.cs [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+using SFA.DAS.RoatpGateway.Web.ViewModels;
+using SFA.DAS.RoatpGateway.Web.Validators;
+
+namespace SFA.DAS.RoatpGateway.Web.Controllers
+{
+    public class RoatpGatewayRegisterChecksController : RoatpGatewayControllerBase<RoatpGatewayRegisterChecksController>
+    {
+        private readonly IGatewayRegisterChecksOrchestrator _orchestrator;
+
+        private const string GatewayViewsLocation = "~/Views/Gateway/pages";
```
Comment (martin-hornsby) [inline src/SFA.DAS.RoatpGateway.Web/Controllers/RoatpGatewayRegisterChecksController.cs:20]: This is in RoatpGatewayControllerBase

File: src/SFA.DAS.RoatpGateway.Web/Controllers/RoatpGatewayPeopleInControlController.cs [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+namespace SFA.DAS.RoatpGateway.Web.Controllers
+{
+
+    public class RoatpGatewayPeopleInControlController : RoatpGatewayControllerBase<RoatpGatewayPeopleInControlController>
+    {
+
+        private readonly IPeopleInControlOrchestrator _orchestrator;
+
+        private const string GatewayViewsLocation = "~/Views/Gateway/pages";
```
Comment (martin-hornsby) [inline src/SFA.DAS.RoatpGateway.Web/Controllers/RoatpGatewayPeopleInControlController.cs:22]: This is in RoatpGatewayControllerBase

File: src/SFA.DAS.RoatpGateway.Web/Infrastructure/ApiClients/RoatpGatewayCriminalComplianceChecksApiClient.cs [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Threading.Tasks;
+
+namespace SFA.DAS.RoatpGateway.Web.Infrastructure.ApiClients
+{
+    public class RoatpGatewayCriminalComplianceChecksApiClient : ApiClientBase<RoatpGatewayCriminalComplianceChecksApiClient>, IRoatpGatewayCriminalComplianceChecksApiClient
+    {
+        private readonly HttpClient _client;
+        private readonly ILogger<RoatpGatewayCriminalComplianceChecksApiClient> _logger;
```
Comment (martin-hornsby) [inline src/SFA.DAS.RoatpGateway.Web/Infrastructure/ApiClients/RoatpGatewayCriminalComplianceChecksApiClient.cs:14]: These are in ApiClientBase

File: src/SFA.DAS.RoatpGateway.Web/Controllers/RoatpGatewayController.cs [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+namespace SFA.DAS.RoatpGateway.Web.Controllers
+{
+    [Authorize(Roles = Roles.RoatpGatewayTeam)]
+    public class RoatpGatewayController : RoatpGatewayControllerBase<RoatpGatewayController>
+    {
+        private readonly IRoatpApplicationApiClient _applyApiClient;
+        private readonly IHttpContextAccessor _contextAccessor;
+        private readonly IGatewayOverviewOrchestrator _orchestrator;
+        private readonly IRoatpGatewayApplicationViewModelValidator _validator;
+        private readonly ILogger<RoatpGatewayController> _logger;
```
Comment (martin-hornsby) [inline src/SFA.DAS.RoatpGateway.Web/Controllers/RoatpGatewayController.cs:25]: This is in RoatpGatewayControllerBase

File: src/SFA.DAS.RoatpGateway.Web/Controllers/RoatpGatewayController.cs [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+using SFA.DAS.RoatpGateway.Web.Services;
+using SFA.DAS.RoatpGateway.Web.Validators;
+using SFA.DAS.RoatpGateway.Web.Domain;
+
+namespace SFA.DAS.RoatpGateway.Web.Controllers
+{
+    [Authorize(Roles = Roles.RoatpGatewayTeam)]
+    public class RoatpGatewayController : RoatpGatewayControllerBase<RoatpGatewayController>
+    {
+        private readonly IRoatpApplicationApiClient _applyApiClient;
```
Comment (martin-hornsby) [inline src/SFA.DAS.RoatpGateway.Web/Controllers/RoatpGatewayController.cs:21]: These are in RoatpGatewayControllerBase

File: src/SFA.DAS.RoatpGateway.Web/Infrastructure/ApiClients/RoatpApplicationApiClient.cs [hunk 13]
```diff
@@ -1,17 +1,233 @@
...
+
+        private async Task<T> Get<T>(string uri)
+        {
+            _client.DefaultRequestHeaders.Authorization =
+                new AuthenticationHeaderValue("Bearer", _tokenService.GetToken(_client.BaseAddress));
+
+            using (var response = await _client.GetAsync(new Uri(uri, UriKind.Relative)))
+            {
+                return await response.Content.ReadAsAsync<T>();
+            }
```
Comment (martin-hornsby) [inline src/SFA.DAS.RoatpGateway.Web/Infrastructure/ApiClients/RoatpApplicationApiClient.cs:229]: This is in RoatpApiClientBase

File: src/SFA.DAS.RoatpGateway.Web/Controllers/RoatpGatewayPeopleInControlController.cs [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+
+        public RoatpGatewayPeopleInControlController(IHttpContextAccessor contextAccessor, IRoatpApplicationApiClient roatpApiClient, ILogger<RoatpGatewayPeopleInControlController> logger, IRoatpGatewayPageValidator validator, IPeopleInControlOrchestrator orchestrator) : base(contextAccessor, roatpApiClient, logger, validator)
+        {
+            {
+                _orchestrator = orchestrator;
+            }
+        }
+
+        [HttpGet("/Roatp/Gateway/{applicationId}/Page/PeopleInControl")]
+        public async Task<IActionResult> GetGatewayPeopleInControlPage(Guid applicationId, string pageId)
```
Comment (martin-hornsby) [inline src/SFA.DAS.RoatpGateway.Web/Controllers/RoatpGatewayPeopleInControlController.cs:32]: pageId is not being used

File: src/SFA.DAS.RoatpGateway.Web/Controllers/RoatpGatewayPeopleInControlController.cs [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+        [HttpPost("/Roatp/Gateway/{applicationId}/Page/PeopleInControl")]
+        public async Task<IActionResult> EvaluatePeopleInControlPage(SubmitGatewayPageAnswerCommand command)
+        {
+            Func<Task<PeopleInControlPageViewModel>> viewModelBuilder = () => _orchestrator.GetPeopleInControlViewModel(new GetPeopleInControlRequest(command.ApplicationId, _contextAccessor.HttpContext.User.UserDisplayName()));
+            return await ValidateAndUpdatePageAnswer(command, viewModelBuilder, $"{GatewayViewsLocation}/PeopleInControl.cshtml");
+        }
+
+
+        [HttpGet("/Roatp/Gateway/{applicationId}/Page/PeopleInControlRisk")]
+        public async Task<IActionResult> GetGatewayPeopleInControlRiskPage(Guid applicationId, string pageId)
```
Comment (martin-hornsby) [inline src/SFA.DAS.RoatpGateway.Web/Controllers/RoatpGatewayPeopleInControlController.cs:48]: pageId is not being used

File: src/SFA.DAS.RoatpGateway.Web/Controllers/RoatpGatewayRegisterChecksController.cs [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+                                                IHttpContextAccessor contextAccessor,
+                                                IRoatpGatewayPageValidator gatewayValidator,
+                                                IGatewayRegisterChecksOrchestrator orchestrator,
+                                                ILogger<RoatpGatewayRegisterChecksController> logger) : base(contextAccessor, applyApiClient, logger, gatewayValidator)
+        {
+            _orchestrator = orchestrator;
+        }
+
+        [HttpGet("/Roatp/Gateway/{applicationId}/Page/Roatp")]
+        public async Task<IActionResult> GetGatewayRoatpPage(Guid applicationId, string pageId)
```
Comment (martin-hornsby) [inline src/SFA.DAS.RoatpGateway.Web/Controllers/RoatpGatewayRegisterChecksController.cs:32]: pageid is not being used

File: src/SFA.DAS.RoatpGateway.Web/Controllers/RoatpGatewayRegisterChecksController.cs [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+
+        [HttpPost("/Roatp/Gateway/{applicationId}/Page/Roatp")]
+        public async Task<IActionResult> EvaluateRoatpPage(SubmitGatewayPageAnswerCommand command)
+        {
+            Func<Task<RoatpPageViewModel>> viewModelBuilder = () => _orchestrator.GetRoatpViewModel(new GetRoatpRequest(command.ApplicationId, _contextAccessor.HttpContext.User.UserDisplayName()));
+            return await ValidateAndUpdatePageAnswer(command, viewModelBuilder, $"{GatewayViewsLocation}/Roatp.cshtml");
+        }
+
+        [HttpGet("/Roatp/Gateway/{applicationId}/Page/Roepao")]
+        public async Task<IActionResult> GetGatewayRoepaoPage(Guid applicationId, string pageId)
```
Comment (martin-hornsby) [inline src/SFA.DAS.RoatpGateway.Web/Controllers/RoatpGatewayRegisterChecksController.cs:47]: pageid is not being used

File: src/SFA.DAS.RoatpGateway.Domain/PaginatedList.cs (hunk 2 above)
Comment (David-B-Read) [reply src/SFA.DAS.RoatpGateway.Domain/PaginatedList.cs:22]: That was lifted and shifted from Admin Service, no idea really

File: src/SFA.DAS.RoatpGateway.Domain/Roatp/OrganisationType.cs (hunk 4 above)
Comment (David-B-Read) [reply src/SFA.DAS.RoatpGateway.Domain/Roatp/OrganisationType.cs:9]: It was copied from Admin Service, where OrganisationType.Unassigned is used in logic, it's not required for Gateway so have removed

**Pull request merged**